File: day2/day2.py
# -*- coding: utf-8 -*-
"""
Created on Mon Dec  2 11:14:25 2019

@author: ttobias
"""

def read(str):
    return list(map(lambda x : int(x), str.split(',')))


def process_instruction(input, pos):
    if input[pos] == 1:
        input[input[pos + 3]] = input[input[pos + 1]] + input[input[pos + 2]]
    elif input[pos] == 2:
        input[input[pos + 3]] = input[input[pos + 1]] * input[input[pos + 2]]
    elif input[pos] == 99:
        return -1
    return pos + 4


def execute(data):
    pos = 0
    while pos != -1:
        pos = process_instruction(data, pos)


def calc(data, noun, verb):
    data_cur = data.copy()
    data_cur[1] = noun
    data_cur[2] = verb

    execute(data_cur)
    return(data_cur[0])


def find_target():
    target = 19690720
    with open('input.txt') as fp:
        file_content = read(fp.read())
    data = read(file_content)

    for noun in range(0,99):
        for verb in range(0,99):
            if calc(data, noun, verb) == target:
                print(noun, verb)
                print(100 * noun + verb)


find_target()

File: fsharp/day2/day2.fs
module Day2

open System.IO

let GetFileContent filename = ((File.ReadAllText filename).Split ',') |> Array.map int

let ProcessInstruction (mem : int[]) pos =
    if mem.[pos] = 1 then
        Array.set mem mem.[pos + 3] (mem.[mem.[pos + 1]] + mem.[mem.[pos + 2]])
        pos + 4
    elif mem.[pos] = 2 then
        Array.set mem mem.[pos + 3] (mem.[mem.[pos + 1]] * mem.[mem.[pos + 2]])
        pos + 4
    elif mem.[pos] = 99 then -1
    else failwith "Unknown instruction"

let Execute memory =
    let mutable pos = 0
    while pos <> -1 do pos <- ProcessInstruction memory pos

let Calc input_memory noun verb =
    let memory = Array.copy input_memory
    Array.set memory 1 noun
    Array.set memory 2 verb
    Execute memory
    memory.[0]

let FindTarget memory target =
    let mutable noun = 0
    let mutable verb = 0
    let mutable found = false

    while noun <= 99 && not found do
        while verb <= 99 && not found do
            found <- (Calc memory noun verb) = target
            if not found then
                verb <- verb + 1
        if not found then
            verb <- 0
            noun <- noun + 1
    if not found then
        failwith "Target not found"
    else
        noun, verb


let solve =
    let data = GetFileContent @"day2\input.txt"
    printfn "Part1: %A" (Calc data 12 2)

    let noun, verb = (FindTarget data 19690720)
    printfn "Part2: %A" (100 * noun + verb)